ostd/src/mm/tlb.rs

```diff
@@ -65,29 +51,75 @@ impl<G: PinCurrentCpu> TlbFlusher<G> {
     /// this CPU or remote CPUs. The flush requests are only performed when
     /// [`Self::dispatch_tlb_flush`] is called.
     pub fn issue_tlb_flush(&self, op: TlbFlushOp) {
-        self.issue_tlb_flush_(op, None);
+        let irq_guard = crate::trap::disable_local();
+        let local_flush_ops = LOCAL_FLUSH_OPS.get_with(&irq_guard);
+        local_flush_ops.borrow_mut().push(op, None);
+    }
+
+    /// Issues a TLB flush request that must happen before dropping the page.
+    ///
+    /// If we need to remove a mapped page from the page table, we can only
+    /// recycle the page after all the relevant TLB entries in all CPUs are
+    /// flushed. Otherwise if the page is recycled for other purposes, the user
+    /// space program can still access the page through the TLB entries. This
+    /// method is designed to be used in such cases.
+    pub fn issue_tlb_flush_with(&self, op: TlbFlushOp, drop_after_flush: Frame<dyn AnyFrameMeta>) {
+        let irq_guard = crate::trap::disable_local();
+        let local_flush_ops = LOCAL_FLUSH_OPS.get_with(&irq_guard);
+        local_flush_ops
+            .borrow_mut()
+            .push(op, Some(drop_after_flush));
     }
 
     /// Dispatches all the pending TLB flush requests.
     ///
-    /// All previous pending requests issued by [`Self::issue_tlb_flush`]
-    /// starts to be processed after this function. But it may not be
-    /// synchronous. Upon the return of this function, the TLB entries may not
-    /// be coherent.
+    /// All previous pending requests issued by [`Self::issue_tlb_flush`] or
+    /// [`Self::issue_tlb_flush_with`] starts to be processed after this
+    /// function. But it may not be synchronous. Upon the return of this
+    /// function, the TLB entries may not be coherent.
     pub fn dispatch_tlb_flush(&mut self) {
-        if !self.need_remote_flush {
+        let irq_guard = crate::trap::disable_local();
+        let local_flush_ops = LOCAL_FLUSH_OPS.get_with(&irq_guard);
+        let mut local_flush_ops = local_flush_ops.borrow_mut();
+
+        if local_flush_ops.is_empty() {
             return;
         }
 
-        for cpu in self.target_cpus.iter() {
-            ACK_REMOTE_FLUSH
-                .get_on_cpu(cpu)
-                .store(false, Ordering::Relaxed);
+        let mut target_cpus = self
+            .target_cpus
+            .map(|s| s.load())
```

Comment:
So, following the above comment, this has to be `load_release` to make the page table changes visible to remote CPUs. But `load(Ordering::Release)` will panic, so we'll have to use something like `fetch_or(0, Ordering::Release)`.